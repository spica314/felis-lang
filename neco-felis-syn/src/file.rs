use crate::{Item, Parse, ParseError, token::Token};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct File {
    items: Vec<Item>,
}

impl File {
    pub fn items(&self) -> &[Item] {
        &self.items
    }
}

impl Parse for File {
    fn parse(tokens: &[Token], i: &mut usize) -> Result<Option<Self>, ParseError> {
        let mut k = *i;

        let mut items = vec![];
        while let Some(item) = Item::parse(tokens, &mut k)? {
            items.push(item);
        }

        let file = File { items };

        *i = k;
        Ok(Some(file))
    }
}

#[cfg(test)]
mod test {
    use crate::{FileIdGenerator, token::Token};

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_inductive_eq() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_eq.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_inductive_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }
}
