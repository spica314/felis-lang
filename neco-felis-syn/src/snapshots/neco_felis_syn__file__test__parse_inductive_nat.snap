---
source: neco-felis-syn/src/file.rs
expression: file
---
File {
    items: [
        Inductive(
            ItemInductive {
                keyword_inductive: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 1,
                    },
                    s: "inductive",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 12,
                    },
                    s: "nat",
                },
                colon: TokenColon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 16,
                    },
                },
                ty: Apply(
                    TermApply {
                        f: Variable(
                            TermVariable {
                                variable: TokenVariable {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 18,
                                    },
                                    s: "Set",
                                },
                            },
                        ),
                        args: [],
                    },
                ),
                brace_l: TokenBraceL {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 22,
                    },
                },
                branches: [
                    ItemInductiveBranch {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 5,
                            },
                            s: "O",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 6,
                            },
                        },
                        ty: Apply(
                            TermApply {
                                f: Variable(
                                    TermVariable {
                                        variable: TokenVariable {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 8,
                                            },
                                            s: "nat",
                                        },
                                    },
                                ),
                                args: [],
                            },
                        ),
                        comma: TokenComma {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 11,
                            },
                        },
                    },
                    ItemInductiveBranch {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 3,
                                column: 5,
                            },
                            s: "S",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 3,
                                column: 6,
                            },
                        },
                        ty: ArrowNodep(
                            TermArrowNodep {
                                from: Variable(
                                    TermVariable {
                                        variable: TokenVariable {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 3,
                                                column: 8,
                                            },
                                            s: "nat",
                                        },
                                    },
                                ),
                                arrow: TokenOperator {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 3,
                                        column: 12,
                                    },
                                    s: "->",
                                },
                                to: Apply(
                                    TermApply {
                                        f: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 3,
                                                        column: 15,
                                                    },
                                                    s: "nat",
                                                },
                                            },
                                        ),
                                        args: [],
                                    },
                                ),
                            },
                        ),
                        comma: TokenComma {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 3,
                                column: 18,
                            },
                        },
                    },
                ],
                brace_r: TokenBraceR {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 4,
                        column: 1,
                    },
                },
            },
        ),
    ],
}
